[FLOC-3495] Cleanup discovered dataset invariant

tomprince opened the pull request:
https://clusterhq.atlassian.net/browse/FLOC-3495

jml commented on flocker/common/algebraic.py:
```diff
@@ -0,0 +1,49 @@
+# -*- test-case-name: flocker.common.test.test_algebraic -*-
+# Copyright Hybrid Logic Ltd.  See LICENSE file for details.
```
ITYM "ClusterHQ"

jml commented on flocker/common/algebraic.py:
```diff
@@ -0,0 +1,49 @@
+__all__ = ["StateInvariant"]
+
+
+class StateInvariant(PClass):
```
It's a bit confusing to use the term "state" (here & elsewhere), because it overlaps with the abstract concept of programming state, and also with states in our state machine, which is where I assume this is derived from. It would be confusing in a different way to use the term "constructor", which is more common in Haskell. 

Looking at https://doc.rust-lang.org/book/enums.html for inspiration, `TaggedUnionInvariant`, with `states` renamed to `tags` and `state` to `tag` might be more clear. 

Alternatively, `EnumInvariant`, either with the "tag" nomenclature, or `variant`.

jml commented on flocker/common/algebraic.py:
```diff
@@ -0,0 +1,49 @@
+
+    :param set allowed_states: Set of allowed states.
+    :param dict expected_attributes: Dictionary mapping attribute names to
+        the set of states that attribute should be present in.
```
I think it would be more clear to invert this dictionary, mapping states to the set of attributes that are allowed, since that's customary for ADTs in most other languages.

jml replied:
As interim steps, I suggest the following:

* Make `allowed_states` and `expected_attributes` both private
* Add helpers: `get_states_for_attribute(self, attribute)`, `get_attributes_for_state(self, state)` (or "tag" or "iter_" or whatever)
* Write the tests and possibly `__call__` in terms of those helpers
* See what's remaining in terms of access to private access and proceed accordingly

At the very least, I think a helper to get valid attributes for a given state would simplify a good chunk of code.

jml commented on flocker/common/algebraic.py:
```diff
@@ -0,0 +1,49 @@
+
+    def __call__(self, value):
+        for attribute, states in self.expected_attributes.items():
+            if (value.state in states) != hasattr(value, attribute):
```
Please add a comment explaining what this conditional is actually checking.

jml commented on flocker/common/algebraic.py:
```diff
@@ -0,0 +1,49 @@
+                            states=', '.join(map("`{0.name}`".format, states)),
+                        )
+                    )
+                return (False, message)
```
:grey_question: Since you're doing early return elsewhere, I'd probably do early return in both branches of the `value.state in states` conditional.

jml commented on flocker/common/test/test_algebraic.py:
```diff
@@ -0,0 +1,117 @@
+# -*- test-case-name: flocker.common.test.test_algebraic -*-
+# Copyright Hybrid Logic Ltd.  See LICENSE file for details.
```
ClusterHQ. The data container people. :wink:

jml commented on flocker/common/algebraic.py:
```diff
@@ -0,0 +1,49 @@
+
+"""
+Invariants for algebraic data types.
+"""
```
A lot of people will have to look up what "algebraic data type" means. It would really help to have an example in the docstring. Please add one.

jml commented on flocker/common/test/test_algebraic.py:
```diff
@@ -0,0 +1,117 @@
+})
+
+
+class MakeStateInvariantTests(SynchronousTestCase):
```
Please add a docstring.

jml commented on flocker/common/test/test_algebraic.py:
```diff
@@ -0,0 +1,117 @@
+            AlgebraicType.__invariant__.expected_attributes
+        ),
+    )
+    def test_extra_attributes(self, args, extra_attribute):
```
Please add a docstring.

jml commented on flocker/common/test/test_algebraic.py:
```diff
@@ -0,0 +1,117 @@
+        ):
+            if attribute == extra_attribute:
+                assume(args['state'] not in states)
+                continue
```
`assume` raises an exception, so there's no need for `continue`

jml commented on flocker/common/test/test_algebraic.py:
```diff
@@ -0,0 +1,117 @@
+                assume(args['state'] not in states)
+                continue
+            if args['state'] not in states:
+                del args[attribute]
```
It took me a while to figure out what this `for` loop was doing. I think it's trying to generate a state with an extra argument that isn't valid in that state.

Rather than doing post-processing in the tests, it's better to have some helpers to generate the right thing. Sketch follows:

```python
states = st.sampled_from(AlgebraicType.__invariant__.allowed_states)
valid_attributes = st.sampled_from(AlgebraicType.__invariant__.expected_attributes)

def args_for_state(state):
    # XXX: This would be simpler if `expected_attributes` were inverted.
    args = {'state': st.just(state)}
    for arg, attrs in AlgebraicType.__invariant__.expected_attributes.items():
        if state in attrs:
            args[arg] = st.booleans()
    return st.fixed_dictionaries(args)

valid_arguments = states.flatmap(args_for_state)

@given(
  args=valid_arguments,
  extra_attribute=valid_attributes,
  extra_arg=st.booleans(),
)
def test_extra_attributes(self, args, extra_attribute, extra_arg):
    assume(extra_attribute not in args)
    args[extra_attribute] = extra_arg
    exc = ...
```

another approach might be to use the new, semi-documented `choices` strategy (http://www.drmaciver.com/2015/11/let-hypothesis-making-your-choices-for-you/)

```python
@given(
  args=valid_arguments,
  extra_arg=st.booleans(),
  choice=st.choices(),
)
def test_extra_attributes(self, args, extra_arg, choice):
    extra_attribute = choice(set(AlgebraicType.__invariant__.allowed_states.keys()) - set(args.keys()))
    args[extra_attribute] = extra_arg
    ...
```

jml commented on flocker/common/test/test_algebraic.py:
```diff
@@ -0,0 +1,117 @@
+            AlgebraicType.__invariant__.expected_attributes
+        ),
+    )
+    def test_missing_attributes(self, args, missing_attribute):
```
Please add a docstring.

jml commented on flocker/common/test/test_algebraic.py:
```diff
@@ -0,0 +1,117 @@
+                assume(args['state'] in states)
+                del args[attribute]
+            if args['state'] not in states:
+                del args[attribute]
```
Similar thing about the loop & conditional. Instead, please generate valid arguments to begin with, and then use `st.choices` to pick something from them to delete.

jml commented on flocker/common/test/test_algebraic.py:
```diff
@@ -0,0 +1,117 @@
+        },
+    )
+
+ARGS_STRATEGEY = st.fixed_dictionaries({
```
Spelling: "strategy", not "strategey".

Although if you make the changes I suggest below (essentially making a strategy to generate valid arguments), then I think this probably won't be necessary any more.

jml commented on flocker/common/test/test_algebraic.py:
```diff
@@ -0,0 +1,117 @@
+        for attribute in (
+            AlgebraicType.__invariant__.expected_attributes.keys()
+        ):
+            del args[attribute]
```
I think this is always going to leave `args` equal to `{"state": state}`, i.e. it will delete all the non-`state` attributes. 

Instead, please either:
 1. pass in an empty dict
 2. generate an arbitrary dictionary of kwargs, and `.map()` it to delete the expected attributes.

Please also add a comment explaining why we're doing this. i.e. that if we pass in invalid arguments, we'll get errors about those rather than about the invalid state.

jml commented on flocker/common/test/test_algebraic.py:
```diff
@@ -0,0 +1,117 @@
+        self.assertIn(
+            'can only be in states',
+            exc.invariant_errors[0],
+        )
```
Please add a test demonstrating that we can construct an `AlgebraicType` with valid arguments.

jml commented on flocker/common/algebraic.py:
```diff
@@ -0,0 +1,49 @@
+    """
+
+    allowed_states = field(set, mandatory=True)
+    expected_attributes = field(dict, mandatory=True)
```
Should these be `pset` and `pmap`? Perhaps even with constraints of their own?

jml commented:
Thanks Tom! 

This is really interesting code, and potentially useful elsewhere. TBH, it took me quite a while to reason through the logic and the tests, so I've made a few suggestions of how they could be made clearer—at least to me. Please resubmit for review after you've made the suggested changes. I'm happy to talk further about the review & comments in real-time (although maybe after the weekend).

Do you have any ideas about pattern matching?

tomprince commented:
I think I have addressed all the review comments.

Having written all the tests, I wonder if they would be clearer if they explicitly tested the cases, instead of using hypothesis (or in addition to using hypothesis). Having a hypothesis strategy for a tagged union is definitely going to be valuable, but using it in the tests of the tagged union code seems like it might be confusing what is being tested between the strategy and the code under test.

jml commented on flocker/common/algebraic.py:
```diff
@@ -0,0 +1,133 @@
+    """
+
+    tag_attribute = field(str, mandatory=True)
+    attributes_for_tag = field(dict, mandatory=True)
```
:question: Should this be a `pmap`? Or even `pmap_field(object, pset)`?

jml commented on flocker/common/algebraic.py:
```diff
@@ -0,0 +1,133 @@
+        """
+        return {
+            attribute
+            for tag, attributes in self.attributes_for_tag.items()
```
:question: Could just use `values()` instead of `items()`

jml commented on flocker/common/algebraic.py:
```diff
@@ -0,0 +1,133 @@
+            for attribute in attributes
+        }
+
+    def __call__(self, value):
```
Please add a docstring.

jml commented on flocker/common/algebraic.py:
```diff
@@ -0,0 +1,133 @@
+                    .format(attr=attribute,
+                            tag_name=self.tag_attribute,
+                            tag=tag.name))
+        return (True, "")
```
:question: I guess if we could get a reliable list of fields that have been set on `value` then we could do this in one loop, iterating over those fields rather than the schema. I don't see a way to do this though.

jml commented on flocker/common/algebraic.py:
```diff
@@ -0,0 +1,133 @@
+                    "`{attr}` can't be specified in {tag_name} `{tag}`"
+                    .format(attr=attribute,
+                            tag_name=self.tag_attribute,
+                            tag=tag.name))
```
It looks like we don't check for fields that have been set that are not valid for any tag? Is this deliberate?

tomprince replied:
Yes. The motivating example, for example, has a bunch of attributes that are required in every state, so I took (and documented) the decision to only care about attributes that should be in some tag (and let pyrsistent's `mandatory` deal with others).

jml commented on flocker/common/algebraic.py:
```diff
@@ -0,0 +1,133 @@
+
+    :param type: Type to generate a strategy for.
+    :param attr_strategies: Mapping of attributes to stratgeies to
+        dgenerate corresponding attributes.
```
Nit: "generate"

jml commented on flocker/common/test/test_algebraic.py:
```diff
@@ -0,0 +1,150 @@
+    'extra': st.booleans(),
+})
+
+ALGEBRAIC_TYPE_ARGUMENTS_STRATGEY = ALGEBRAIC_TYPE_STRATEGY.map(
```
Nit: `STRATEGY`, not `STRATGEY`

jml commented on flocker/common/test/test_algebraic.py:
```diff
@@ -0,0 +1,150 @@
+            - invariant.attributes_for_tag[state]
+        )
+        assume(extra_attributes)
+        extra_attribute = choice(sorted(extra_attributes))
```
:question: Any particular reason for `sorted` here?

tomprince replied:
Yes. `choice` doesn't accept `set`s and since hypothesis keeps examples around, I thought it would be good to have a stable ordering to pick from.

jml commented:
Thanks!

Only a few minor things to do. Please address those & merge when the tests pass.

As for testing strategy, it might be a good idea to _add_ a couple of explicit non-hypothesis test cases, if only as extra documentation. I'll leave it up to you as to whether to do this, and to decide on whether a follow-up review is needed. 

There's probably scope for future work here, esp. with factoring out some top-level functions to extract data & metadata about tagged unions: what's the tag attribute? what's the value of the tag attribute? what are valid tags? However, presently I think all that's YAGNI.

tomprince commented:
> As for testing strategy, it might be a good idea to add a couple of explicit non-hypothesis test cases, if only as extra documentation.

Revisiting this, I realized that I could add explicit hypothesis examples to test, that has about the same effect.
